ultravox/data/datasets.py

```diff
@@ -550,77 +549,48 @@ def __len__(self):
         return self._length
 
 
-class StopStrategy(str, enum.Enum):
-    FIRST_EXHAUSTED = "FIRST_EXHAUSTED"
-    LAST_EXHAUSTED = "LAST_EXHAUSTED"
-    NEVER_STOP = "NEVER_STOP"
-
-
 class InterleaveDataset(SizedIterableDataset):
-    """Interleaves multiple IterableDataset objects based on normalized weights."""
+    """Interleaves multiple SizedIterableDataset objects based on normalized weights."""
 
     def __init__(
         self,
         datasets: Sequence[SizedIterableDataset],
         weights: Optional[Sequence[float]] = None,
-        stop_strategy: StopStrategy = StopStrategy.LAST_EXHAUSTED,
-        seed: Optional[int] = 42,
-        static: bool = False,
     ) -> None:
         """
         Args:
             datasets: A list of SizedIterableDataset objects.
-            weights: A list of weights for each dataset.
-            stop_strategy: Strategy for stopping iteration.
+            weights: An optional list of dataset weights, i.e., the number of times it should be repeated.
             seed: Optional seed for reproducibility.
-            static: If true, the datasets are interleaved in a static order with equal weights.
         """
         self._datasets = datasets
-        self._rng = np.random.default_rng(seed)
-        self._static = static
-        self._stop_strategy = stop_strategy
-
-        if weights is None:
+        if weights is not None:
+            assert len(weights) == len(datasets)
+        else:
             weights = [1.0] * len(datasets)
-        total_weight = sum(weights)
-        self._normalized_probs = [w / total_weight for w in weights]
+        self._weighted_samples = [int(w * len(d)) for w, d in zip(weights, datasets)]
+        self._total_samples = sum(self._weighted_samples)
 
     def __iter__(self):
-        # If no datasets are provided, return an empty iterator
-        if not self._datasets:
-            return
-
-        iters = [iter(ds) for ds in self._datasets]
-        exhausted = [False] * len(iters)
-
-        if self._static:
-            static_iter = itertools.cycle(range(len(self._datasets)))
-
-        while True:
-            if self._static:
-                iter_index = next(static_iter)
-            else:
-                iter_index = self._rng.choice(len(iters), p=self._normalized_probs)
-
+        ds_iters = [iter(ds) for ds in self._datasets]
+        ds_pos = [0] * len(ds_iters)
+        # Find the iterator that is least far along and vend from it.
+        for i in range(self._total_samples):
```

Comment:
This for loop can cause issues.

I believe this method is gonna work fine if the weights are whole numbers, but with fractions and specially `weight < 1` I'm not so sure:

`ds_iters` will get re-started when the for loop ends and:
1. Unless we reshuffle the dataset, you're always going to get the same subset. That means if you set `weight=0.1`, you're never going to see 90% of the data.
2. Even with reshuffling, since the iterator gets reset before you do a full epoch, you might see some samples multiple times while some samples haven't been seen yet.

Reply:
See this for item 1:
```python
>> ds = datasets.load_dataset('librispeech_asr', split="validation.clean", streaming=True)
>> ds = ds.shuffle()
>>> next(iter(ds))
{'file': '6345-64257-0015.flac', 'audio': {'path': '6345-64257-0015.flac', 'array': array([-0.00195312, -0.00149536, -0.00149536, ...,  0.00015259,
        0.00054932,  0.00143433]), 'sampling_rate': 16000}, 'text': 'I KNOW IT AND THERE IS NO WAKING', 'speaker_id': 6345, 'chapter_id': 64257, 'id': '6345-64257-0015'}
>>> next(iter(ds))
{'file': '6345-64257-0015.flac', 'audio': {'path': '6345-64257-0015.flac', 'array': array([-0.00195312, -0.00149536, -0.00149536, ...,  0.00015259,
        0.00054932,  0.00143433]), 'sampling_rate': 16000}, 'text': 'I KNOW IT AND THERE IS NO WAKING', 'speaker_id': 6345, 'chapter_id': 64257, 'id': '6345-64257-0015'}
 ```

Reply:
Agreed. In a sense the weight is really just the number of "epochs" to run for the dataset, so from that view the behavior matches expectations (for both 0.1 and 1.1). Is there a situation where you'd see the same 0.1 of samples multiple times?

If needed, one easy thing we could do would be to add a random offset to each dataset. For 1.0 you would get the same samples just in a unique order, but for 0.1 you'd get a random 10% slice.

Reply:
Not sure I understood what you said:

> Is there a situation where you'd see the same 0.1 of samples multiple times?

Is that a question of do we ever want this? The answer is I think no. I don't see a real advantage in having that. Maybe just for testing :?

> one easy thing we could do would be to add a random offset to each dataset

That's kinda similar to what I suggested with shuffling and so the issue would be the same: with `weight=0.1`, if you do 10 epochs, you're gonna see some samples multiple times and some samples 0 times.